R/MsExperiment-db.R

```diff
@@ -0,0 +1,166 @@
+    if (!length(spectra(x))) return()
+    if (inherits(spectra(x)@backend, "MsBackendSql")) {
+        be <- spectra(x)@backend
+        if (inherits(be, "MsBackendOfflineSql")) {
+            flags <- be@flags
+            be@flags <- 2L # fix to ensure getting a writebale connection
+            con <- dbconn(be)
+            be@flags <- flags
+            on.exit(dbDisconnect(con))
+        } else con <- dbconn(be)
+        if (dbExistsTable(con, "sample_data")) {
+            message("Sample data already available in the database. ",
+                    "Will overwrite.")
+            dbRemoveTable(con, "sample_data")
+        }
+        if (dbExistsTable(con, "sample_to_msms_spectrum"))
+            dbRemoveTable(con, "sample_to_msms_spectrum")
+        sdata <- as.data.frame(sampleData(x))
+        if (any(colnames(sdata) == "sample_id_"))
+            warning("Replacing already present column \"sample_id_\" ",
+                    "in 'sampleData'.")
+        sdata$sample_id_ <- seq_len(nrow(sdata))
+        dbWriteTable(con, name = "sample_data", sdata)
+        sdata_links <- x@sampleDataLinks[["spectra"]]
+        if (length(sdata_links)) {
+            sdata_links <- as.data.frame(sdata_links)
+            colnames(sdata_links) <- c("sample_index", "spectrum_index")
+            dbWriteTable(con, name = "sample_to_msms_spectrum", sdata_links)
+        }
+    } else stop("The object's 'Spectra' object does not use a ",
+                "'MsBackendSql' backend.")
+}
+
+#' @title Get sample data from a MsBackendSql database
+#'
+#' @description
+#'
+#' Retrieve sample data and sample to spectra links from a MsBackendSql
+#' database, if present. This requires that the `Spectra` `x` uses a
+#' `MsBackendSql` or `MsBackendOfflineSql` backend and that the database
+#' contains database tables *sample_data* and eventually
+#' *sample_to_msms_spectrum*.
+#'
+#' @param x `Spectra`
+#'
+#' @author Johannes Rainer
+#'
+#' @return information from the database or an empty array.
+#'
+#' @importFrom DBI dbGetQuery dbListTables
+#'
+#' @importMethodsFrom BiocGenerics dbconn
+#'
+#' @noRd
+.db_get_sample_data <- function(x) {
+    if (inherits(x@backend, "MsBackendSql")) {
+        con <- dbconn(x@backend)
+        if (inherits(x@backend, "MsBackendOfflineSql"))
+            on.exit(dbDisconnect(con))
+        if (any(dbListTables(con) == "sample_data"))
+            dbGetQuery(con, "select * from sample_data")
+        else data.frame()
+    } else data.frame()
+}
+.db_get_sample_spectra_link <- function(x) {
+    if (inherits(x@backend, "MsBackendSql")) {
+        con <- dbconn(x@backend)
+        if (inherits(x@backend, "MsBackendOfflineSql"))
+            on.exit(dbDisconnect(con))
+        if (any(dbListTables(con) == "sample_to_msms_spectrum"))
+            unname(as.matrix(
+                dbGetQuery(con, "select * from sample_to_msms_spectrum")))
+        else matrix(ncol = 2, nrow = 0)
+    } else matrix(ncol = 2, nrow = 0)
```

Comment:
actually I am a bit confused with the annotation "sample_to_msms_spectrum". is it only for MS2 ? If it is, then I have a hard time understanding the code could you explain what the different steps do here

Reply:
yes, the link matrix is a two column matrix with the indices (IDs/primary keys) to link samples and spectra - in the form:


| sample | spectra |
|--|--|
| 1 | 1 |
| 1 | 2 |
| 1 | 3 |
| 1 | 4 |
| 2 | 5 |
| 2 | 6 |
| ... | ... |

this allows us also to have n:m relationships (although at present we use only 1:n, i.e. one sample has many spectra, but each spectra is associated only to one sample).

And yes, agree, the name might be a bit misleading. In fact, the database has tables called *sample_data* (for the sample annotations), *msms_spectrum* (for the spectra variables, one row for each spectrum) and *msms_spectrum_peak* (for the peak variables, i.e. the m/z and intensity values). So, I chose the name *sample_to_msms_spectrum* as database table name here to indicate that this table provides the mapping from sample to spectrum. The name *msms_spectrum* is a bit misleading indeed because that table contains just MS spectrum data (MS1, MS2, ...). The database scheme is the same that we use(d) in `CompoundDb`, and there it's mostly just MS2 spectra in that table. I just kept the same database layout so we can re-use code across packages.

Hope this explains it.

Reply:
so the *msms_spectrum* tables contains both MS1 and MS2 data ? So there is a column in this table that precise the msLevel ?

Reply:
yes, exactly. that table contains all the data from `spectraData()` of a `Spectra` object, i.e. a column for each of the `spectraVariables()` (including `msLevel`). The database layout actually represents the structure within a `Spectra` - it's one table for the `spectraVariables` and one for the `peakVariables`. I wanted to keep it as simple as possible.

Reply:
Okay that clear now thanks 👍